core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java

```diff
@@ -0,0 +1,205 @@
+   private static abstract class DefaultComposedCompletionStage<R> extends CompletableFuture<R>
+         implements ComposedCompletionStage<R>, BiConsumer<Object, Throwable> {
+      private static final AtomicIntegerFieldUpdater<DefaultComposedCompletionStage> remainingUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(DefaultComposedCompletionStage.class, "remaining");
+
+      @SuppressWarnings({"unused"})
+      private volatile int remaining;
+      private volatile boolean frozen = false;
+      private volatile Throwable throwable;
+
+      @Override
+      public void accept(Object o, Throwable t) {
+         if (t != null) {
+            throwable = t;
+         }
+         if (remainingUpdater.decrementAndGet(this) == 0 && frozen) {
+            complete();
+         }
+      }
+
+      @Override
+      final public ComposedCompletionStage<R> dependsOn(CompletionStage<?> stage) {
+         Objects.requireNonNull(stage);
+         if (frozen) {
+            throw new IllegalStateException();
+         }
+         // We only depend upon it if the stage wasn't complete
+         if (!isComplete(stage)) {
+            remainingUpdater.incrementAndGet(this);
+            stage.whenComplete(this);
+         }
+         return this;
+      }
+
+      @Override
+      final public ComposedCompletionStage<R> freeze() {
+         frozen = true;
+         if (remainingUpdater.get(this) == 0) {
+            complete();
+         }
+         return this;
+      }
+
+      private void complete() {
+         Throwable t = throwable;
+         if (t != null) {
+            completeExceptionally(t);
+         } else {
+            complete(getValue());
+         }
+      }
+
+      abstract R getValue();
+   }
+
+   /**
+    * Returns if the provided {@link CompletionStage} has already completed normally, that is not due to an exception.
+    * @param stage stage to check
+    * @return if the stage is completed normally
+    */
+   public static boolean isComplete(CompletionStage<?> stage) {
+      CompletableFuture<?> future = stage.toCompletableFuture();
+      return future.isDone() && !future.isCompletedExceptionally();
+   }
+
+   /**
+    * Returns the result value when complete, or throws an (unchecked) exception if completed exceptionally.
+    * To better conform with the use of common functional forms, if a computation involved in the completion of this
+    * CompletionStage threw an exception, this method throws an (unchecked) CompletionException with the underlying
+    * exception as its cause.
+    * @param stage stage to wait on
+    * @param <R> the type in the stage
+    * @return the result value
+    * @throws CompletionException if this stage completed exceptionally or a completion computation threw an exception
+    */
+   public static <R> R join(CompletionStage<R> stage) {
+      return stage.toCompletableFuture().join();
```

Comment:
Hrmm, it returns the value though, which is join. Maybe you meant something else?

Reply:
I did mean await: this method should use `CompletableFutures.await()` instead of `CompletableFuture.join()` to avoid the busy wait, and since they do the same thing they should have the same name.

I know you already have other methods named `await` that don't return the value, but having the same method name mean different things would be confusing, so we should either change the name of `CompletableFutures.await()` or the names of the methods you're adding here.

Reply:
The other difference is that the other await is throwing exceptions, which is one thing I was trying to avoid (another difference of join vs get). I was trying to keep the semantics like join is on CompletableFuture myself.

I will change it to invoke the `CompletableFutures.await` though. And I will also rename or remove the other await methods, not sure which yet.